src/framework/results/experiment_result.hpp

```diff
@@ -0,0 +1,111 @@
+/**
+ * This code is part of Qiskit.
+ *
+ * (C) Copyright IBM 2018, 2019.
+ *
+ * This code is licensed under the Apache License, Version 2.0. You may
+ * obtain a copy of this license in the LICENSE.txt file in the root directory
+ * of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Any modifications or derivative works of this code must retain this
+ * copyright notice, and modified files need to carry a notice indicating
+ * that they have been altered from the originals.
+ */
+
+#ifndef _aer_framework_results_experiment_result_hpp_
+#define _aer_framework_results_experiment_result_hpp_
+
+#include "framework/results/experiment_data.hpp"
+
+namespace AER {
+
+//============================================================================
+// Result container for Qiskit-Aer
+//============================================================================
+
+struct ExperimentResult {
+public:
+
+  // Status 
+  enum class Status {empty, completed, error};
+
+  // Experiment data
+  ExperimentData data;
+  uint_t shots;
+  uint_t seed;
+  double time_taken;
+
+  // Success and status
+  Status status = Status::empty;
+  std::string message; // error message
+
+  // Metadata
+  json_t header;
+  json_t metadata;
+
+  // Clear all metadata for given key
+  void clear_metadata(const std::string &key);
+  
+  // Append metadata for a given key.
+  // This assumes the metadata value is a dictionary and appends
+  // any new values
+  template <typename T>
+  void add_metadata(const std::string &key, const T &data);
+
+  // Serialize engine data to JSON
+  json_t json() const;
+};
+
+
+//------------------------------------------------------------------------------
+// Add metadata
+//------------------------------------------------------------------------------
+template <typename T>
+void ExperimentResult::add_metadata(const std::string &key, const T &meta) {
```

Comment:
This is exactly the same method found in `Result::add_metadata()`... inheriting from 
`Result` will remove the need to duplicate code.